text/0000-proc-macros.md

```diff
@@ -0,0 +1,452 @@
+a record of macro expansion is maintained (and it will make life easier for
+tools. Again, we might optimise internally).
+
+A token is a single lexical element, for example, a numeric literal, a word
+(which could be an identifier or keyword), a string literal, or a comment.
+
+A token stream is a sequence of tokens, e.g., `a b c;` is a stream of four
+tokens - `['a', 'b', 'c', ';'']`.
+
+A token tree is a tree structure where each leaf node is a token and each
+interior node is a token stream. I.e., a token stream which can contain nested
+token streams. A token tree can be delimited, e.g., `a (b c);` will give
+`TT(None, ['a', TT(Some('()'), ['b', 'c'], ';'']))`. An undelimited token tree
+is useful for grouping tokens due to expansion, without representation in the
+source code. That could be used for unsafety hygiene, or to affect precedence
+and parsing without affecting scoping. They also replace the interpolated AST
+tokens currently in the compiler.
+
+In code:
+
+```
+// We might optimise this representation
+pub struct TokenStream(Vec<TokenTree>);
+
+// A borrowed TokenStream
+pub struct TokenSlice<'a>(&'a [TokenTree]);
+
+// A token or token tree.
+pub struct TokenTree {
+    pub kind: TokenKind,
+    pub span: Span,
+    pub hygiene: HygieneObject,
+}
+
+pub enum TokenKind {
+    Sequence(Delimiter, TokenStream),
+
+    // The content of the comment can be found from the span.
+    Comment(CommentKind),
+
+    // Symbol is the string contents, not including delimiters. It would be nice
+    // to avoid an allocation in the common case that the string is in the
+    // source code. We might be able to use `&'Codemap str` or something.
+    // `Option<usize> is for the count of `#`s if the string is a raw string. If
+    // the string is not raw, then it will be `None`.
+    String(Symbol, Option<usize>, StringKind),
+
+    // char literal, span includes the `'` delimiters.
+    Char(char),
+
+    // These tokens are treated specially since they are used for macro
+    // expansion or delimiting items.
+    Exclamation,  // `!`
+    Dollar,       // `$`
+    // Not actually sure if we need this or if semicolons can be treated like
+    // other punctuation.
+    Semicolon,    // `;`
+    Eof,          // Do we need this?
+
+    // Word is defined by Unicode Standard Annex 31 -
+    // [Unicode Identifier and Pattern Syntax](http://unicode.org/reports/tr31/)
+    Word(Symbol),
+    Punctuation(char),
+}
+
+pub enum Delimiter {
+    None,
+    // { }
+    Brace,
+    // ( )
+    Parenthesis,
+    // [ ]
+    Bracket,
+}
+
+pub enum CommentKind {
+    Regular,
+    InnerDoc,
+    OuterDoc,
+}
+
+pub enum StringKind {
+    Regular,
+    Byte,
+}
+
+// A Symbol is a possibly-interned string.
+pub struct Symbol { ... }
+```
+
+### Open question: `Punctuation(char)` and multi-char operators.
+
+Rust has many compound operators, e.g., `<<`. It's not clear how best to deal
```

Comment:
Sorry, that was too vague. Initially I was thinking we could use an `&str` but since we probably don't want to thread a lifetime, you could imagine instead that we just pass enough info that you can ask the `MacroContext` for the precise string:

```
impl MacroContext {
    fn token_str(&self, token: &Token) -> &str;
}
```